k2/csrc/pinned_context.cu

```diff
@@ -0,0 +1,343 @@
+      if (err == cudaErrorNotReady) break;
+
+      if (err != cudaSuccess) return err;
+
+      err = cudaEventDestroy(event);
+      if (err != cudaSuccess) return err;
+
+      Block &block = blocks_.at(e.second);
+      --block.event_count;
+
+      if (block.event_count == 0 && !block.allocated) available_.insert(block);
+
+      cuda_events_.pop_front();
+    }
+    return cudaSuccess;
+  }
+
+ private:
+  // It contains all blocks allocated by Malloc.
+  std::unordered_map<void *, Block> blocks_;
+
+  using Compare = bool (*)(const BlockSize &, const BlockSize &);
+  // It contains all free blocks **sorted** by block size in increasing order.
+  std::set<BlockSize, Compare> available_;
+
+  // outstanding cuda events
+  std::deque<std::pair<cudaEvent_t, void *>> cuda_events_;
+
+  // to protect `blocks_`, `available_` and `cuda_events_` being accessed
+  // from multiple threads
+  std::mutex mutex_;
+};
+
+static PinnedAllocator *GetPinnedAllocator() {
+  static std::once_flag init_flag;
+  static PinnedAllocator *allocator = nullptr;
+
+  std::call_once(init_flag, []() {
+    // it is never freed.
+    allocator = new PinnedAllocator;
+  });
+
+  return allocator;
+}
+
+}  // namespace
+
+class PinnedContext : public Context {
+ public:
+  PinnedContext() { allocator_ = GetPinnedAllocator(); }
+
+  DeviceType GetDeviceType() const override { return kCpu; }
+
+  void *Allocate(std::size_t bytes, void **deleter_context) override {
+    void *p = nullptr;
+    cudaError_t err = allocator_->Malloc(bytes, &p);
+    if (deleter_context != nullptr) *deleter_context = nullptr;
+    return p;
+  }
+
+  void Deallocate(void *data, void *deleter_context) override {
+    (void)deleter_context;
+    allocator_->Free(data);
+  }
+
+  bool IsCompatible(const Context &other) const override {
+    return other.GetDeviceType() == kCpu;
+  }
+
+  void CopyDataTo(size_t num_bytes, const void *src, ContextPtr dst_context,
+                  void *dst) override {
+    DeviceType device_type = dst_context->GetDeviceType();
+    switch (device_type) {
+      case kCpu:
+        // we assume that src and dst do not overlap
+        memcpy(dst, src, num_bytes);
+        break;
+      case kCuda: {
+        cudaStream_t stream = dst_context->GetCudaStream();
+        cudaError_t ret = cudaMemcpyAsync(dst, src, num_bytes,
+                                          cudaMemcpyHostToDevice, stream);
+        K2_CHECK_CUDA_ERROR(ret);
+
+        allocator_->RecordEvent(stream, const_cast<void *>(src));
+        break;
+      }
+      default:
+        K2_LOG(FATAL) << "Unsupported device type: " << device_type;
+        break;
+    }
+  }
+
+ private:
+  PinnedAllocator *allocator_;  // NOT owned here
+};
+
+ContextPtr GetPinnedContext() {
+  static std::once_flag has_cuda_init_flag;
+  static bool has_cuda = false;
+
+  std::call_once(has_cuda_init_flag, []() {
+    int32_t count = 0;
+    cudaError_t err = cudaGetDeviceCount(&count);
```

Comment:
So what does it do in `torch::cuda::is_available` below, I just feel it may be duplicate with below `InitHasCuda` stuff?

Reply:
pinned_context does not depend on PyTorch. The implementation is similar to `torch::cuda::is_available`.